# **Question Answer Models**

## **Import Packages and Data**

In [1]:
! pip install transformers
! pip install sentence_transformers
! pip install accelerate

In [2]:
import time
from tqdm.auto import tqdm
import pickle
import accelerate

import math
import numpy as np
import pandas as pd

from sklearn.metrics.pairwise import cosine_similarity

from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords as nltk_stopwords

import torch
import transformers
from transformers import BertForQuestionAnswering, BertTokenizer
from transformers import LlamaForCausalLM, LlamaTokenizer, AutoTokenizer
from sentence_transformers import SentenceTransformer

In [3]:
from huggingface_hub import login
login(token='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')

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [ ]:
df_QA = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Externship/qa_merged_clean.csv',
                    parse_dates=True)

In [ ]:
df_QA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 987122 entries, 0 to 987121
Data columns (total 23 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             987122 non-null  int64  
 1   id_q                   987122 non-null  int64  
 2   owner_user_id_q        973748 non-null  float64
 3   creation_date_q        987122 non-null  object 
 4   score_q                987122 non-null  int64  
 5   title                  987122 non-null  object 
 6   body_q                 987122 non-null  object 
 7   body_normalized_q      987120 non-null  object 
 8   title_normalized       987122 non-null  object 
 9   body_with_sentences_q  987122 non-null  object 
 10  title_with_sentences   987122 non-null  object 
 11  creation_year_q        987122 non-null  int64  
 12  id_a                   987122 non-null  float64
 13  owner_user_id_a        981755 non-null  float64
 14  creation_date_a        987122 non-nu

In [ ]:
print(df_QA.isna().sum())

Unnamed: 0                   0
id_q                         0
owner_user_id_q          13374
creation_date_q              0
score_q                      0
title                        0
body_q                       0
body_normalized_q            2
title_normalized             0
body_with_sentences_q        0
title_with_sentences         0
creation_year_q              0
id_a                         0
owner_user_id_a           5367
creation_date_a              0
parent_id                    0
score_a                      0
body_a                       0
body_normalized_a            7
body_with_sentences_a        5
creation_year_a              0
answer_length                0
question_length              0
dtype: int64


In [ ]:
df_QA=df_QA.reset_index(drop=True)

## **QA Models**

### Subset data for questions with answers and scores above 0

In [ ]:
# Subset for Q & A with positive scores
df_QA = df_QA[(df_QA['score_a'] >= 0) & (df_QA['score_q'] >= 0)]

print('Answer Score Descriptives')
print(df_QA['score_a'].describe())
print()
print('Question Score Descriptives')
print(df_QA['score_q'].describe())

Answer Score Descriptives
count    913099.000000
mean          3.239337
std          22.090040
min           0.000000
25%           0.000000
50%           1.000000
75%           3.000000
max        8384.000000
Name: score_a, dtype: float64

Question Score Descriptives
count    913099.000000
mean          7.769584
std          65.424836
min           0.000000
25%           0.000000
50%           1.000000
75%           3.000000
max        5524.000000
Name: score_q, dtype: float64


All question and answer scores now have a minimum of 0.

There are 913,099 remaining Q/A pairs where both question and answer have scores > 0 and every question has at least one answer.

### BERT

In [ ]:
bert_model = BertForQuestionAnswering.from_pretrained('bert-large-uncased-whole-word-masking-finetuned-squad')
bert_tokenizer = BertTokenizer.from_pretrained('bert-large-uncased-whole-word-masking-finetuned-squad')

Some weights of the model checkpoint at bert-large-uncased-whole-word-masking-finetuned-squad were not used when initializing BertForQuestionAnswering: ['bert.pooler.dense.weight', 'bert.pooler.dense.bias']
- This IS expected if you are initializing BertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


#### Model with specific answer as context
result: got an answer!

In [ ]:
# Choose row 10
display(df_QA.loc[10])

Unnamed: 0                                                              10
id_q                                                                   535
owner_user_id_q                                                      154.0
creation_date_q                                  2008-08-02 18:43:54+00:00
score_q                                                                 40
title                    Continuous Integration System for a Python Cod...
body_q                   <p>I'm starting work on a hobby project with a...
body_normalized_q        i'm starting work on a hobby project with a py...
title_normalized         continuous integration system for a python cod...
body_with_sentences_q    i'm starting work on a hobby project with a py...
title_with_sentences     continuous integration system for a python cod...
creation_year_q                                                       2008
id_a                                                               61746.0
owner_user_id_a          

In [ ]:
# Ask a relevant question for row 10
question = 'how to do hobby project?'
answer_text = df_QA.loc[10, 'body_a']

In [ ]:
# BERT QA with row 10
start_time = time.time()

input_ids = bert_tokenizer.encode(question, answer_text)

attention_mask = [1] * len(input_ids)

output = bert_model(torch.tensor([input_ids]), attention_mask=torch.tensor([attention_mask]))

start_index = torch.argmax(output[0][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])
end_index = torch.argmax(output[1][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])

answer = bert_tokenizer.decode(input_ids[start_index:end_index + 1], skip_special_tokens=True)

end_time = time.time()

computation_time = end_time - start_time

print(f'Answer: {answer}')
print(f'Computation Time: {computation_time:.2f} seconds')

Answer: build it yourself
Computation Time: 4.39 seconds


#### Find similar questions

Get embedding for first 10,000 Qs

In [ ]:
df_QA = df_QA.reset_index(drop=True)

In [ ]:
# Get embeddings for data set questions
start_time = time.time()

questions = df_QA.loc[0:10000, 'body_with_sentences_q']

sent_model = SentenceTransformer('bert-base-nli-mean-tokens')

ques_embeddings = sent_model.encode(questions)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question Embeddings Shape: {ques_embeddings.shape}')
print(f'Computation Time: {computation_time:.2f} seconds')

Question Embeddings Shape: (10001, 768)
Computation Time: 1181.21 seconds


In [ ]:
with open('/Users/kellyshreeve/Desktop/ques_embeddings', 'wb') as file:
    pickle.dump(ques_embeddings, file)

In [ ]:
file = open('/Users/kellyshreeve/desktop/ques_embeddings', 'rb')
pickled_embeddings = pickle.load(file)

In [ ]:
pickled_embeddings.shape

(10001, 768)

Find similar questions with cosine similarity

In [ ]:
# Use cosine distance to find similar questions
start_time = time.time()

new_question = 'What is pandas?'

model = SentenceTransformer('bert-base-nli-mean-tokens')

new_question_embeddings = model.encode(new_question)

similarity_scores = cosine_similarity([new_question_embeddings],
                                       ques_embeddings)

best_index = np.argmin(similarity_scores)

end_time = time.time()

computation_time = end_time - start_time

best_question = df_QA.loc[best_index, 'body_with_sentences_q']
best_answer = df_QA.loc[best_index, 'body_with_sentences_a']

print('Question Posed:')
print(new_question)
print()
print('Question:')
print()
print(best_question)
print()
print('Answer:')
print()
print(best_answer)
print()
print(f'Best Index: {best_index}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Question:

so when playing with the development i can just set settings.debug to true and if an error occures i can see it nicely formatted with good stack trace and request information. but on kind of production site i'd rather use debug false and show visitors some standard error page with information that i'm working on fixing this bug at this moment br at the same time i'd like to have some way of logging all those information stack trace and request info to a file on my server so i can just output it to my console and watch errors scroll email the log to me every hour or something like this. what logging solutions would you recomend for a django site that would meet those simple requirements i have the application running as fcgi server and i'm using apache web server as frontend although thinking of going to lighttpd .

Answer:

well when debug false django will automatically mail a full traceback of any error to each person listed in the admins setting which gets you notificatio

Find answer to question from answer to similar question

In [ ]:
# Find answer from most relevant question
start_time = time.time()

new_question = 'What is pandas?'

input_ids = bert_tokenizer.encode(new_question, df_QA.loc[5822, 'body_a'])

attention_mask = [1] * len(input_ids)

output = bert_model(torch.tensor([input_ids]), attention_mask=torch.tensor([attention_mask]))

start_index = torch.argmax(output[0][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])
end_index = torch.argmax(output[1][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])

answer = bert_tokenizer.decode(input_ids[start_index:end_index + 1], skip_special_tokens=True)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question: {new_question}')
print()
print(f'Answer: {answer}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Answer: 

Computation Time: 1.97 seconds


BERT did not find an answer in the given answer text.

Check that BERT QA works for a more relevant question

In [ ]:
# Find an answer from a more relevant question
start_time = time.time()

relevant_question = 'How to use django?'

input_ids = bert_tokenizer.encode(relevant_question, df_QA.loc[5822, 'body_a'])

attention_mask = [1] * len(input_ids)

output = bert_model(torch.tensor([input_ids]), attention_mask=torch.tensor([attention_mask]))

start_index = torch.argmax(output[0][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])
end_index = torch.argmax(output[1][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])

answer = bert_tokenizer.decode(input_ids[start_index:end_index + 1], skip_special_tokens=True)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question: {relevant_question}')
print(f'Answer: {answer}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Answer: django will automatically mail a full traceback of any error to each person listed in the < code > admins < / code > setting

Computation Time: 2.19 seconds


While not a great answer, BERT did find an answer in the answer text.

In [ ]:
# Similar question function
def find_similar_question(question, df, question_column):
    new_question_embeddings = model.encode(new_question)

    similarity_scores = cosine_similarity([new_question_embeddings],
                                        ques_embeddings)

    best_index = np.argmin(similarity_scores)

    print(f'Posed Question: {question}')
    print(f'Most Similar Question: {df.loc[best_index, question_column]})

# Try new questions
find_similar_question('What is python?', df_QA, 'body_with_sentences_q')

### Find similar answers

Get embeddings for first 10,000 answers

In [ ]:
# Get embeddings for data set answers
start_time = time.time()

sentence_model = SentenceTransformer('bert-base-nli-mean-tokens')

answers = df_QA.loc[0:10000, 'body_with_sentences_a']

answer_embeddings = sentence_model.encode(answers)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question Embeddings Shape: {ques_embeddings.shape}')
print(f'Computation Time: {computation_time:.2f} seconds')

Question Embeddings Shape: (10001, 768)
Computation Time: 1125.99 seconds


Find answers similar to question

In [ ]:
# Use cosine distance to find answer similar to question
start_time = time.time()

new_question = 'What is pandas?'

model = SentenceTransformer('bert-base-nli-mean-tokens')

new_question_embeddings = sentence_model.encode(new_question)

similarity_scores = cosine_similarity([new_question_embeddings],
                                       answer_embeddings)

best_index = np.argmin(similarity_scores)

end_time = time.time()

computation_time = end_time - start_time

best_question = df_QA.loc[best_index, 'body_with_sentences_q']
best_answer = df_QA.loc[best_index, 'body_with_sentences_a']

print('Question Posed:')
print()
print(new_question)
print()
print('Question:')
print()
print(best_question)
print()
print('Answer:')
print()
print(best_answer)
print()
print(f'Best Index: {best_index}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Question Posed:

What is pandas?

Question:

i recently discovered the notify extension in mercurial which allows me quickly send out emails whenever i push changes but i'm tty sure i'm still missing out on a lot of functionality which could make my life a lot easier. ul li notify extension http www.selenic.com mercurial wiki index.cgi notifyextension rel nofollow http www.selenic.com mercurial wiki index.cgi notifyextension li ul which mercurial hook or combination of interoperating hooks is the most useful for working in a loosely connected team please add links to non standard parts you use and or add the hook or a description how to set it up so others can easily use it.

Answer:

i really enjoy what i did with my custom hook. i have it post a message to my campfire account campfire is a group based app . it worked out really well. because i had my clients in there and it could show him my progress.

Best Index: 1419

Computation Time: 1.18 seconds


Also not a relevant answer. Difficult to know if it would work better with more embeddings.

### Llama 2

In [4]:
start = time.time()

model_id = "meta-llama/Llama-2-7b-chat-hf"

llama_tokenizer = AutoTokenizer.from_pretrained(
    model_id)

pipeline = transformers.pipeline(
    'text-generation',
    model="meta-llama/Llama-2-7b-chat-hf",
    torch_dtype=torch.float16,
    device_map='auto'
)

end = time.time()

print(f'Computation Time: {end - start}')

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [7]:
start = time.time()

sequences = pipeline(
    'I liked "Breaking Bad" and "Band of Brothers". Do you have any recommendations of other shows I might like?\n',
    do_sample=True,
    top_k=10,
    num_return_sequences=3,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: I liked "Breaking Bad" and "Band of Brothers". Do you have any recommendations of other shows I might like?

I am a big fan of crime dramas and historical dramas. I enjoy shows with complex characters and intricate storylines. I also enjoy shows with a strong sense of atmosphere and setting.

I am open to trying new things, so if you have any recommendations, please let me know!
Result: I liked "Breaking Bad" and "Band of Brothers". Do you have any recommendations of other shows I might like?

Answer: Yes, I can definitely recommend some other shows that you might enjoy based on your interest in "Breaking Bad" and "Band of Brothers". Here are a few suggestions:

1. "The Wire" - This critically acclaimed HBO series explores the drug trade in Baltimore from multiple perspectives, including law enforcement, drug dealers, and politicians. Like "Breaking Bad," it features complex characters and a gripping storyline.
2. "Sons of Anarchy" - This FX series follows the lives of a close-

In [8]:
start = time.time()

sequences = pipeline(
    "I'm working in python. Do you know what the pandas packages is?",
    do_sample=True,
    top_k=10,
    num_return_sequences=1,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: I'm working in python. Do you know what the pandas packages is?

Comment: Yes, I'm familiar with the pandas package in Python. It's a powerful library for data manipulation and analysis. Pandas provides efficient data structures and operations for working with structured data, including tabular data such as spreadsheets and SQL tables.

It allows you to perform various data manipulation tasks such as reading and writing data to various file formats, merging and reshaping data, and performing statistical operations on data.

Some common tasks that pandas can help you with include:

1. Reading and writing data to various file formats, such as CSV, Excel, and SQL.
2. Merging and reshaping data from multiple sources.
3. Performing statistical operations on data, such as aggregating and filtering data.
4. Creating and manipulating data visualizations, such as charts and plots.

Pandas is
Result: I'm working in python. Do you know what the pandas packages is?

Answer: Yes, I'm famili

In [9]:
start = time.time()

sequences = pipeline(
    'Do you know how I add a column to a dataframe in Pandas?',
    do_sample=True,
    top_k=10,
    num_return_sequences=3,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: Do you know how I add a column to a dataframe in Pandas?

Answer: Yes, you can add a column to a Pandas DataFrame in several ways:

1. Using the `loc` method:
```
df = df.loc[:, df.columns.get_value('new_column_name')]
```
This will add a new column to the DataFrame with the name `new_column_name`.

2. Using the `iloc` method:
```
df = df.iloc[:, df.columns.get_value('new_column_name')]
```
This will add a new column to the DataFrame at the index position specified by `new_column_name`.

3. Using the `add_column` method:
```
df = df.add_column(new_column_name, value)
```
This will add a new column
Result: Do you know how I add a column to a dataframe in Pandas?

Answer: Yes, you can add a column to a Pandas DataFrame using the `loc` method and the `assign` method.

Here is an example of how to add a column to a DataFrame using the `loc` method:
```
# create a sample DataFrame
df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})

# add a new column to the DataFrame using the loc 

In [10]:
start = time.time()

sequences = pipeline(
    'How do you make a baked potato?',
    do_sample=True,
    top_k=10,
    num_return_sequences=1,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: How do you make a baked potato?

To make a baked potato, you will need the following ingredients:

* 1-2 large potatoes (depending on size)
* 1/4 cup vegetable oil
* Salt and pepper to taste
* Optional toppings: cheese, sour cream, chives, bacon bits, etc.

Instructions:

1. Preheat your oven to 400°F (200°C).
2. Scrub the potatoes clean and pat them dry with a paper towel.
3. Use a fork to poke a few holes in each potato. This will allow steam to escape while the potatoes are baking.
4. Rub the potatoes with vegetable oil and sprinkle with salt and pepper.
5. Place the potatoes directly on the middle ra
Result: How do you make a baked potato?
To make a baked potato, preheat your oven to 400°F (200°C). Scrub a potato clean and dry it thoroughly with a paper towel. Use a fork to poke a few holes in the potato, then rub it with a little bit of oil and sprinkle with salt. Place the potato directly on the middle rack of the oven and bake for 45 to 60 minutes, or until it's cooked t

### **User Function**

### QA Function v.1 - initial

In [ ]:
def question_answer():
    start_time = time.time()
    
    # Take in user question
    posed_question = input('Question:')
    
    # Load data frame
    df_QA = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Externship/qa_merged_clean.csv')
    
    # Load question embeddings
    file = open('/Users/kellyshreeve/desktop/embeddings', 'rb')
    ques_embeddings = pickle.load(file)
    
    # Initiate Sentence Model
    sent_model = SentenceTransformer('bert-base-nli-mean-tokens')

    # Get new question embeddings
    new_question_embeddings = sent_model.encode(posed_question)

    # Find most similar question index
    similarity_scores = cosine_similarity([new_question_embeddings],
                                       ques_embeddings)

    best_index = np.argmax(similarity_scores)
    
    # Extract similar question and answer text from df
    best_question = df_QA.loc[best_index, 'body_with_sentences_q']
    best_answer = df_QA.loc[best_index, 'body_with_sentences_a']
    
    end_time = time.time()
    
    computation_time = end_time - start_time
    
    # Print results
    print(f'Posed Question: {posed_question}')
    print(f'Similar Question: {best_question}')
    print(f'Similar Answer: {best_answer}')
    print(f'Embeddings Shape: {ques_embeddings.shape}')
    print(f'Computation Time: {computation_time}')

In [ ]:
question_answer()

KeyboardInterrupt: 

### QA Function v.2 - faster

Pull data, embedding, and model load out of function.

In [ ]:
# Load all data and sentence model
start = time.time()

# Load data frame
df_QA = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Externship/qa_merged_clean.csv')
    
# Load question embeddings
file = open('/Users/kellyshreeve/desktop/ques_embeddings', 'rb')
ques_embeddings = pickle.load(file)

# Initiate Sentence Model
sent_model = SentenceTransformer('bert-base-nli-mean-tokens')

end = time.time()

print(f'Compuation Time: {end - start}')

Compuation Time: 45.44052767753601


In [ ]:
# Function to get user input and embeddings and return
# similar Q/A. Does not load data or embeddings.

def question_answer_v2():
    start_time = time.time()
    
    # Take in user question
    posed_question = input('Question:')

    # Get new question embeddings
    new_question_embeddings = sent_model.encode(posed_question)

    # Find most similar question index
    similarity_scores = cosine_similarity([new_question_embeddings],
                                       ques_embeddings)

    best_index = np.argmax(similarity_scores)
    
    # Extract similar question and answer text from df
    best_question = df_QA.loc[best_index, 'body_with_sentences_q']
    best_answer = df_QA.loc[best_index, 'body_with_sentences_a']
    
    end_time = time.time()
    
    computation_time = end_time - start_time
    
    # Print results
    print(f'Posed Question: {posed_question}')
    print()
    print(f'Similar Question: {best_question}')
    print()
    print(f'Similar Answer: {best_answer}')
    print()
    print(f'Embeddings Shape: {ques_embeddings.shape}')
    print()
    print(f'Computation Time: {computation_time}')

In [ ]:
question_answer()

Posed Question: What is django?

Similar Question: how do i go about specifying and using an enum in a django model

Similar Answer: from the https docs.djangoproject.com en dev ref models fields django.db.models.field.choices rel nofollow django documentation maybechoice 'y' 'yes' 'n' 'no' 'u' 'unknown' and you define a charfield in your model married models.charfield max_len h choices maybechoice you can do the same with integer fields if you don't like to have letters in your db. in that case rewrite your choices maybechoice 'yes' 'no' 'unknown'

Embeddings Shape: (10001, 768)

Computation Time: 4.076124906539917


In [ ]:
question_answer()

Posed Question: How to add a column in Pandas?

Similar Question: how do you change the size of figure drawn with matplotlib

Similar Answer: the following seems to work from pylab import rcparams rcparams['figure.figsize'] this makes the figure's width inches and its height b inches b . the figure class then uses this as the default value for one of its arguments.

Embeddings Shape: (10001, 768)

Computation Time: 12.959703922271729


In [ ]:
question_answer()

Posed Question: How to find a full path to a font?

Similar Question: does anyone know how to do this i need to add a header of the form value value

Similar Answer: as the question is phrased it's hard to guess what the intention or even the intended semantics is. for setting headers try the following import soappy headers soappy.types.headertype headers.value value or [...] headers.foo value headers.bar value

Embeddings Shape: (10001, 768)

Computation Time: 7.04338002204895


In [ ]:
question_answer()

Posed Question: How to find a full path to a font in photoshop javascript?

Similar Question: is there any python module for rendering a html page with javascript and get back a dom object i want to parse a page which generates almost all of its content using javascript.

Similar Answer: only way i know to accomplish this would be to drive real browser for example using http selenium rc.openqa.org rel nofollow selenium rc .

Embeddings Shape: (10001, 768)

Computation Time: 14.047891855239868


### QA Function v.3 - normalize question text

In [ ]:
# Load all data and sentence model
start = time.time()

# Load data frame
df_QA = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Externship/qa_merged_clean.csv')
    
# Load question embeddings
file = open('/Users/kellyshreeve/desktop/ques_embeddings', 'rb')
ques_embeddings = pickle.load(file)

# Initiate Sentence Model
sent_model = SentenceTransformer('bert-base-nli-mean-tokens')

end = time.time()

print(f'Compuation Time: {end - start}')

Compuation Time: 42.505502223968506


In [ ]:
def normalize_with_sentences(text):
    text = text.lower()
    text = text.replace('<p>', ' ')
    text = text.replace('</p>', ' ')
    text = text.replace('\n', ' ')
    text = text.replace('<a', ' ')
    text = text.replace('</a>', ' ')
    text = text.replace('href=', ' ')
    text = text.replace('</code', ' ')
    text = text.replace('</pre>', ' ')
    text = text.replace('<code>', ' ')
    text = text.replace('jpeg', ' ')
    text = text.replace('jpg', ' ')
    text = text.replace('pre', ' ')
    text = text.replace('pdf', ' ')
    text = text.replace('gt', ' ')
    text = re.sub(r"[^a-zA-z'.]", ' ', text)
    text = text.split()
    text = " ".join(text)
    
    return text

In [ ]:
# Function to get user input and embeddings and return
# similar Q/A. Does not load data or embeddings.
# Normalizes question text

def question_answer_v3():
    start_time = time.time()
    
    # Take in user question
    posed_question = input('Question:')
    
    # Normalize question
    posed_quesiton = normalize_with_sentences(posed_question)

    # Get new question embeddings
    new_question_embeddings = sent_model.encode(posed_question)

    # Find most similar question index
    similarity_scores = cosine_similarity([new_question_embeddings],
                                       ques_embeddings)

    best_index = np.argmax(similarity_scores)
    
    # Extract similar question and answer text from df
    best_question = df_QA.loc[best_index, 'body_with_sentences_q']
    best_answer = df_QA.loc[best_index, 'body_with_sentences_a']
    
    end_time = time.time()
    
    computation_time = end_time - start_time
    
    # Print results
    print(f'Posed Question: {posed_question}')
    print()
    print(f'Similar Question: {best_question}')
    print()
    print(f'Similar Answer: {best_answer}')
    print()
    print(f'Embeddings Shape: {ques_embeddings.shape}')
    print()
    print(f'Computation Time: {computation_time}')

In [ ]:
question_answer_v3()

Posed Question: What is python?

Similar Question: how do you create a weak reference to an object in python

Similar Answer: import weakref class object ... pass ... o object r weakref.ref o if the reference is still active r will be o otherwise none do_something_with_o r see the http docs.python.org lib module weakref.html wearkref module docs for more details. you can also use weakref.proxy to create an object that proxies o. will throw referenceerror if used when the referent is no longer referenced.

Embeddings Shape: (10001, 768)

Computation Time: 5.623437166213989


In [ ]:
question_answer_v3()

Posed Question: What is pandas?

Similar Question: how do i turn a python program into an .egg file

Similar Answer: http peak.telecommunity.com devcenter setuptools setuptools is the software that creates http peak.telecommunity.com devcenter pythoneggs .egg files . it's an extension of the http docs.python.org lib module distutils.html distutils package in the standard library. the process involves creating a setup.py file then python setup.py bdist_egg creates an .egg package.

Embeddings Shape: (10001, 768)

Computation Time: 9.889105796813965


In [ ]:
question_answer()

Posed Question: What is Django?

Similar Question: how do i go about specifying and using an enum in a django model

Similar Answer: from the https docs.djangoproject.com en dev ref models fields django.db.models.field.choices rel nofollow django documentation maybechoice 'y' 'yes' 'n' 'no' 'u' 'unknown' and you define a charfield in your model married models.charfield max_len h choices maybechoice you can do the same with integer fields if you don't like to have letters in your db. in that case rewrite your choices maybechoice 'yes' 'no' 'unknown'

Embeddings Shape: (10001, 768)

Computation Time: 8.851870059967041


In [ ]:
question_answer()

Posed Question: How to find the full path to a font?

Similar Question: does anyone know how to do this i need to add a header of the form value value

Similar Answer: as the question is phrased it's hard to guess what the intention or even the intended semantics is. for setting headers try the following import soappy headers soappy.types.headertype headers.value value or [...] headers.foo value headers.bar value

Embeddings Shape: (10001, 768)

Computation Time: 8.934815883636475
